src/dsql/DdlNodes.epp

```diff
@@ -11978,32 +11981,35 @@ void GrantRevokeNode::checkGrantorCanGrantRelation(thread_db* tdbb, jrd_tra* tra
 
 	FOR(REQUEST_HANDLE request TRANSACTION_HANDLE transaction)
 		PRV IN RDB$USER_PRIVILEGES WITH
-			PRV.RDB$USER = UPPERCASE(grantor) AND
-			PRV.RDB$USER_TYPE = obj_user AND
+			((PRV.RDB$USER = UPPERCASE(grantor) AND
+				PRV.RDB$USER_TYPE = obj_user) OR (PRV.RDB$USER_TYPE = obj_sql_role)) AND
 			PRV.RDB$RELATION_NAME = relationName.c_str() AND
 			PRV.RDB$OBJECT_TYPE = obj_relation AND
 			PRV.RDB$PRIVILEGE = privilege
 	{
+		if ( (PRV.RDB$USER_TYPE == obj_sql_role) && !attachment->att_user->roleInUse(tdbb, PRV.RDB$USER))
+			continue;
+
+		const bool hasGrantOption = !PRV.RDB$GRANT_OPTION.NULL && PRV.RDB$GRANT_OPTION;
 		if (PRV.RDB$FIELD_NAME.NULL)
 		{
-			if (PRV.RDB$GRANT_OPTION.NULL || !PRV.RDB$GRANT_OPTION)
-				goRel = 0;
-			else if (goRel)
+			if (goRel == -1)
+				goRel = hasGrantOption ? WITH_GRANT_OPTION : 0;
+			else if ((goRel == 0) && hasGrantOption)
 				goRel = WITH_GRANT_OPTION;
 		}
 		else
 		{
-			if (PRV.RDB$GRANT_OPTION.NULL || !PRV.RDB$GRANT_OPTION)
+			if (fieldName.hasData() && fieldName == PRV.RDB$FIELD_NAME)
 			{
-				if (fieldName.hasData() && fieldName == PRV.RDB$FIELD_NAME)
-					goFld = 0;
-			}
-			else
-			{
-				if (fieldName.hasData() && fieldName == PRV.RDB$FIELD_NAME)
-					goFld = 1;
+				if (goFld == -1)
+					goFld = hasGrantOption ? WITH_GRANT_OPTION : 0;
+				else if ((goFld == 0) && hasGrantOption)
+					goFld = WITH_GRANT_OPTION;
 			}
 		}
```

Comment:
I do not like an idea of leaving FOR/END_FOR loop using break. This keeps busy resources used by request.

Reply:
AFAIU destructor of AutoCacheRequest should be called both for exit after the last iteration of loop and for some intermediate one. I see no any special handles after END_FOR. Are there releasing resources after the last call EXE_receive? Moreover I see "returns" from loop in functions: isItSqlRole, getIndexRelationName, getTriggerRelationName.